### PR TITLE
Enable caching stored procedure results -> REST (GET) GraphQL (Queries)

### LINKED ISSUE #2299: [Enhancement] Cache - Stored Procedure Support
Per #2253

### PR DESCRIPTION
**Pull request by seantleonard**
## Why make this change?
- Closes #2299 
- Enabling DAB's cache only worked for generic table/view queries (REST: GET, GraphQL: queries). Stored procedure results were always fetched from the database.
- This change enables caching for stored procedure results when the stored procedure is configured as such:
```json
"entities": {
  "MyStoredProcedureEntity": {
    "source": {
      "type": "stored-procedure",
      "object": "dbo.spExample"
    },
    "graphql": {
      "operation": "query"
    },
    "rest": {
      "methods": [ "GET" ]
    },
    "cache": {
      "enabled": true,
      "ttl-seconds": 5
    },
    "permissions": [
      {
        "role": "anonymous",
        "actions": [
          "execute"
        ]
      }
    ]
  }
```

## What is this change?

- Updates DAB's SqlQueryEngine function  `private async Task<JsonDocument?> ExecuteAsync(SqlExecuteStructure structure, string dataSourceName)` to utilize DAB's cache service. 
- The code now checks whether caching is enabled globally:
  - `RuntimeConfig.Cache.Enabled` is **true**
  - `RuntimeConfig.DataSource.Options.SetSessionContext` is **false**
- The cache service is invoked with type `JsonArray` : `_cache.GetOrSetAsync<JsonArray?>(...)`
## How was this tested?

- [ ] Integration Tests
- [x] Unit Tests: tests that cache invocation with type `JsonArray` succeeds. This covers how stored procedures are executed in the query engine. 

## Sample Request(s)
Dab config: use example entity at top of this description.
Example stored procedure:

```tsql
create procedure [dbo].[stp_Dummy]
@mandatoryParam int,
@optionalParam int = 10
as
select 
    @mandatoryParam as mandatoryParam,
    @optionalParam as optionalParam
GO
```

REST
```https
GET http://localhost:5000/api/Dummy?mandatoryParam=123&optionalParam=2345
```

GraphQL
```graphql
query mySP{
  executeDummy {
    mandatoryParam
    optionalParam
  }
}
```

### CODE REVIEW

File: src/Core/Resolvers/SqlQueryEngine.cs [hunk 1]
```diff
@@ -353,11 +353,48 @@ public object ResolveList(JsonElement array, IObjectField fieldSchema, ref IMeta
...
+                bool entityCacheEnabled = runtimeConfig.Entities[structure.EntityName].IsCachingEnabled;
+
+                // Stored procedures do not support nor honor runtime config defined
+                // authorization policies. Here, DAB only checks that the entity has
+                // caching enabled and doesn't check for database policies. This explicitly
+                // differs from how the cache works for non-stored procedure requests.
+                if (entityCacheEnabled)
+                {
+                    DatabaseQueryMetadata queryMetadata = new(
+                        queryText: queryString,
```
Comment (abhishekkumams) [inline src/Core/Resolvers/SqlQueryEngine.cs:377]: parameters are not properly aligned.

**Comment by abhishekkumams** [comment]
how would it work, when caching is disabled globally but enabled at entity level or vice-versa?

File: src/Core/Resolvers/SqlQueryEngine.cs [hunk 2]
```diff
@@ -353,11 +353,48 @@ public object ResolveList(JsonElement array, IObjectField fieldSchema, ref IMeta
...
+                            sqltext: queryString,
+                            parameters: structure.Parameters,
+                            dataReaderHandler: queryExecutor.GetJsonArrayAsync,
+                            httpContext: _httpContextAccessor.HttpContext!,
+                            args: null,
+                            dataSourceName: dataSourceName),
+                        queryMetadata,
+                        runtimeConfig.GetEntityCacheEntryTtl(entityName: structure.EntityName));
+
+                    byte[] jsonBytes = JsonSerializer.SerializeToUtf8Bytes(result);
```
[the remark this answers is not in the document]
Comment (seantleonard) [reply src/Core/Resolvers/SqlQueryEngine.cs:392]: good catch. Now following how other code ensures that null result isn't serialized and just return null jsondoc.

**Comment by seantleonard** [comment]
> how would it work, when caching is disabled globally but enabled at entity level or vice-versa?

For caching:
1. Global cache must be enabled (with session context turned off)
2. Entity cache must be turned on + no db policy configured.

image

**Comment by Aniruddh25** [comment]
Would this cache results even if the stored proc has a mutation operation in it? If thats the case, in a cache hit, will the stores proc not be executed? So, for cache hits, we will NOT execute the mutation operation.. Is the responsibility to ensure they enable cache only for READ-ONLY stored procs with our customer developers?

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs [hunk 3]
```diff
@@ -252,6 +253,90 @@ await Assert.ThrowsExceptionAsync<DataApiBuilderException>(
                 message: "Expected an exception to be thrown.");
         }
 
+        /// <summary>
+        /// Tests DAB's cache service invocation when the type is JsonArray.
+        /// JsonArray aligns with the type used for executing stored procedures against
+        /// MSSQL databases.
+        /// This test validates that the cache service returns the expected a database response
```
Comment (Aniruddh25) [inline src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:260]:
```suggestion
        /// This test validates that the cache service returns the expected database response
```

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs [hunk 4]
```diff
@@ -252,6 +253,90 @@ await Assert.ThrowsExceptionAsync<DataApiBuilderException>(
...
+            Dictionary<string, DbConnectionParam> parameters = new()
+            {
+                {"param1", new DbConnectionParam(value: "param1Value") }
+            };
+
+            DatabaseQueryMetadata queryMetadata = new(queryText: "select c.name from c", dataSource: "dataSource1", queryParameters: parameters);
+            DabCacheService dabCache = CreateDabCacheService(cache);
+
+            // Act
+            int cacheEntryTtl = 1;
```
Comment (Aniruddh25) [inline src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:286]:
```suggestion
            int cacheEntryTtlInSeconds = 1;
```

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs [hunk 5]
```diff
@@ -252,6 +253,90 @@ await Assert.ThrowsExceptionAsync<DataApiBuilderException>(
...
+            // First call. Cache miss
+            _ = await dabCache.GetOrSetAsync<JsonArray>(executeQueryAsync: mockExecuteQuery.Object, queryMetadata: queryMetadata, cacheEntryTtl: cacheEntryTtl);
+
+            // Act
+            JsonArray? result = await dabCache.GetOrSetAsync<JsonArray>(executeQueryAsync: mockExecuteQuery.Object, queryMetadata: queryMetadata, cacheEntryTtl: cacheEntryTtl);
+
+            // Assert
+            Assert.AreEqual(expected: true, actual: mockExecuteQuery.Invocations.Count is 1, message: ERROR_UNEXPECTED_INVOCATIONS);
+            Assert.IsFalse(mockExecuteQuery.Invocations.Count > 1, message: "Expected a cache hit, but observed cache misses.");
+            Assert.AreEqual(expected: true, actual: mockExecuteQuery.Invocations.Count is 1, message: ERROR_UNEXPECTED_INVOCATIONS);
```
Comment (Aniruddh25) [inline src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:336]: redundant assert here on line 336?

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs [hunk 6]
```diff
@@ -252,6 +253,90 @@ await Assert.ThrowsExceptionAsync<DataApiBuilderException>(
...
 
+        /// <summary>
+        /// Tests DAB's cache service invocation when the type is JsonArray.
+        /// JsonArray aligns with the type used for executing stored procedures against
+        /// MSSQL databases.
+        /// This test validates that the cache service returns the expected a database response
+        /// because the cache is empty and the factory method is expected to be called.
+        /// </summary>
+        [TestMethod]
+        public async Task JsonArray_CacheServiceInvokation_CacheEmpty_ReturnsFactoryResult()
```
Comment (Aniruddh25) [inline src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:264]:
```suggestion
        public async Task JsonArray_CacheServiceInvocation_CacheEmpty_ReturnsFactoryResult()
```

**Review by Aniruddh25 — APPROVED** [review]
Looks good, left a few questions and suggestions

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs (hunk 5 above)
Comment (seantleonard) [reply src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:336]: i saw one other instance of this. Removing

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs (hunk 3 above)
Comment (seantleonard) [reply src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs]: fixed

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs (hunk 5 above)
Comment (seantleonard) [reply src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs:335]: fixed

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs (hunk 6 above)
Comment (seantleonard) [reply src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs]: fixed

File: src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs (hunk 4 above)
Comment (seantleonard) [reply src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs]: fixed

**Comment by seantleonard** [comment]
@Aniruddh25 

> Would this cache results even if the stored proc has a mutation operation in it?

No, this code is within the QueryEngine so Stored Procedures configured as mutations would not enter this code path.

**Review by abhishekkumams — APPROVED** [review]

**Pull request merged**